### PR TITLE
Line Following with Agent-Based Decision Making

### PR DESCRIPTION
**Pull request by akrambzeo**
Implemented an agent for decision making in line following, added a trajectory buffer and networks, and created a plot file for analysis of the results.

### CODE REVIEW

File: webots/controllers/RL_Supervisor/agent.py [hunk 1]
```diff
@@ -0,0 +1,544 @@
...
+                estimated_pos = 0
+
+            # If it last read to the right of center, return the max. value.
+            else:
+                estimated_pos = (max_sensors - 1) * sensor_max_value
+
+        else:
+            # Check to avoid division by zero..
+            if denominator != 0:
+                estimated_pos = numerator / denominator
```
Comment (BlueAndi) [inline webots/controllers/RL_Supervisor/agent.py:320]: Be aware that the result will be a double and not an integer! To use integer division, use //

File: webots/controllers/RL_Supervisor/agent.py [hunk 2]
```diff
@@ -0,0 +1,544 @@
...
+MAX_SENSOR_VALUE = 1000
+MIN_STD_DEV = 0.1  # Minimum standard deviation
+STD_DEV_FACTOR = 0.99995  # Discounter standard deviation factor
+
+################################################################################
+# Classes
+################################################################################
+
+
+class Agent:  # pylint: disable=too-many-instance-attributes
```
Comment (BlueAndi) [inline webots/controllers/RL_Supervisor/agent.py:76]: The agent is huge, think about splitting it in logical parts and export them to separate classes.

File: webots/controllers/RL_Supervisor/trajectory_buffer.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+        """Remove transitions from the replay buffer."""
+
+        self.__states = []
+        self.__probs = []
+        self.__vals = []
+        self.__actions = []
+        self.__rewards = []
+        self.__dones = []
+
+    def is_memory_full(self):
```
Comment (BlueAndi) [inline webots/controllers/RL_Supervisor/trajectory_buffer.py:133]: Sometimes the return type is described and sometimes not, like here. Please define it consistently and check that for all methods which return something.

**Review by BlueAndi — CHANGES_REQUESTED** [review]
Please increase number of PRs and keep logical changes small.

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 4]
```diff
@@ -117,7 +117,10 @@ class ReadyState : public IState
     };
 
     /** Duration of the selected mode in ms. This is the maximum time to select a mode. */
-    static const uint32_t mode_selected_period = 200U;
+    static const uint32_t mode_selected_period = 100;
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/ReadyState.h:120]: If the user does not press a button, the system automatically switches to training mode after a timer expires to ensure automatic switching.

File: lib/APPReinforcementLearning/src/DrivingState.cpp [hunk 5]
```diff
@@ -79,9 +79,15 @@ void DrivingState::exit()
     m_observationTimer.stop();
 }
 
-void DrivingState::setTargetSpeeds(int16_t leftMotor, int16_t rightMotor)
+void DrivingState::setTargetSpeeds(int16_t leftSpeed, int16_t rightSpeed)
 {
-    DifferentialDrive::getInstance().setLinearSpeed(leftMotor, rightMotor);
+    DifferentialDrive& diffDrive       = DifferentialDrive::getInstance();
+    const int16_t      MAX_MOTOR_SPEED = diffDrive.getMaxMotorSpeed();
+    const int16_t      MIN_MOTOR_SPEED = 0;
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/DrivingState.cpp:86]: Yes, this is intended to minimise complexity in the training.

File: lib/APPReinforcementLearning/src/App.h [hunk 6]
```diff
@@ -141,6 +142,9 @@ class App
     /** Ensue that the mode is only sent once */
     bool m_modeSelectionSent;
 
+    /** check if the data has been sent */
+    bool m_dataSent;
```
Comment (BlueAndi) [inline lib/APPReinforcementLearning/src/App.h:146]: Why still using this member instead of using a method local variable in the loop()?

File: lib/APPReinforcementLearning/src/ReadyState.h [hunk 7]
```diff
@@ -117,7 +117,10 @@ class ReadyState : public IState
     };
 
     /** Duration of the selected mode in ms. This is the maximum time to select a mode. */
-    static const uint32_t mode_selected_period = 200U;
+    static const uint32_t MODE_SELECTED_PERIOD = 100U;
```
Comment (BlueAndi) [inline lib/APPReinforcementLearning/src/ReadyState.h:120]: Now it might work with the timeout value, but I don't think you can change between the modes by the buttons. How shall a human press a button within 150ms before the transitition will take place? Please check this manually!

BTW why does the getSelectedMode() returns uint8_t instead of Mode?

**Review by BlueAndi — CHANGES_REQUESTED** [review]

File: lib/APPReinforcementLearning/src/ReadyState.h (hunk 7 above)
Comment (akrambzeo) [reply lib/APPReinforcementLearning/src/ReadyState.h]: The getSelectedMode() function originally returned uint8_t instead of the mode enum because this was a bug in the implementation. I fixed this bug by changing the return value to the mode enum.

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 8]
```diff
@@ -61,107 +63,126 @@
...
+
+        # Detect Start/Stop Line before Finish Trajectories
+        if (self.start_stop_line_detected is True) and (self.steps < MIN_NUMBER_OF_STEPS):
+            sensor_data = list(sensor_data)
+            sensor_data[SENSOR_ID_MOST_LEFT] = 0
+            sensor_data[SENSOR_ID_MOST_RIGHT] = 0
+            self.start_stop_line_detected = False
+
+        # sequence stop criterion debounce no line detection and start/stop line detected
+        if self.__no_line_detection_count >= 30 or (
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:124]: Use parenthesis correctly

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 9]
```diff
@@ -61,107 +63,126 @@
...
+                    self.__agent.action,
+                    self.__agent.adjusted_log_prob,
+                    self.__agent.value,
+                    reward,
+                    self.__agent.done,
+                )
+            self.last_sensor_data = sensor_data
+
+        # Sends the motor speeds to the robot.
+        if self.__agent.done is False and self.__agent.state == "READY":
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:154]: Use parenthesis. can "READY" be a constant or enum?

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 10]
```diff
@@ -177,6 +198,8 @@ def process(self):
 
 
 # pylint: disable=duplicate-code
+# pylint: disable=too-many-branches
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:201]: optimiza instead of suppressing warnings

File: webots/controllers/RL_Supervisor/trajectory_buffer.py [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+################################################################################
+# Imports
+################################################################################
+
+import numpy as np  # pylint: disable=import-error
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/trajectory_buffer.py:31]: See comment on imports

**Review by gabryelreyes — CHANGES_REQUESTED** [review]
For every python function that returns something, please try to use type-hinting for the return values on the function definition. This helps better follow and understand the code

File: webots/controllers/RL_Supervisor/networks.py [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+# Variables
+################################################################################
+
+
+################################################################################
+# Classes
+################################################################################
+
+
+class Models:
```
Comment (hoeftjch) [inline webots/controllers/RL_Supervisor/networks.py:45]: Consider deriving from tf.keras.Model to adhere to Keras architecture and match it's design principles.
https://www.tensorflow.org/api_docs/python/tf/keras/Model

File: webots/controllers/RL_Supervisor/trajectory_buffer.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+# Variables
+################################################################################
+
+
+################################################################################
+# Classes
+################################################################################
+
+
+class Memory:  # pylint: disable=too-many-instance-attributes
```
Comment (hoeftjch) [inline webots/controllers/RL_Supervisor/trajectory_buffer.py:43]: Consider using the tensorflow dataset classes. This should improve performance due to less conversions from python lists to numpy arrays to tensorflow tensors.
https://www.tensorflow.org/guide/data_performance

File: webots/controllers/RL_Supervisor/rl_supervisor.py [hunk 14]
```diff
@@ -61,107 +63,126 @@
...
-            )
+            # receive a -1 punishment if the robot leaves the line
+            if self.__no_line_detection_count > 0:
+                reward = -1
+            else:
+                reward = self.__agent.determine_reward(sensor_data)
+
+            # Start storage The data after the second received sensor data
+            if self.last_sensor_data is not None:
+                normalized_sensor_data = self.__agent.normalize_sensor_data(self.last_sensor_data)
```
Comment (hoeftjch) [inline webots/controllers/RL_Supervisor/rl_supervisor.py:142]: Perform preprocessing and transformations at a later stage. This function should be used for data aquisition only. If more input data than just line sensor data is used or the preprocesing operation takes more time than the sampling.

File: webots/controllers/RL_Supervisor/agent.py [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+        """
+        Checks if the sequence has ended and performs updates.
+
+        Parameters
+        ----------
+            robot_node: The Robot interface
+        """
+
+        # Checks whether the sequence has ended if it is set to Training mode.
+        if (self.train_mode is True) and (
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py:254]: The parenthesis will end in line 256.

File: webots/controllers/RL_Supervisor/agent.py [hunk 16]
```diff
@@ -0,0 +1,507 @@
...
+            # Failed to send data. Appends the data to unsent_data List
+            if self.data_sent is False:
+                self.unsent_data.append(("SPEED_SET", motorcontrol))
+
+            self.reinitialize(robot_node)
+            self.data_sent = self.__serialmux.send_data("CMD", cmd_payload)
+
+            # Failed to send data. Appends the data to unsent_data List
+            if self.data_sent is False:
+                self.unsent_data.append(("CMD", cmd_payload))
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py:287]: In every timestep, I check in the supervisor whether any data has not been sent. If that's the case, I use the retry_unsent_data function in the supervisor at line 174 to resend the data.

File: webots/controllers/RL_Supervisor/agent.py [hunk 17]
```diff
@@ -0,0 +1,507 @@
...
+    def perform_training(self):
+        """Runs the training process."""
+
+        if self.__current_batch is None:
+
+            # Grab sample from memory
+            self.__current_batch = self.__memory.generate_batches()
+
+        # Perform training with mini batches.
+        if self.__training_index < len(self.__current_batch[-1]):
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py:443]: current_batch[-1] is the list of batches, and its length tells  how many batches are available, training_index tracks which batch  currently using.  The condition ensures that  only accessing valid batch indices during training, preventing errors

File: webots/controllers/RL_Supervisor/agent.py [hunk 18]
```diff
@@ -0,0 +1,507 @@
...
+        Parameters
+        ----------
+            sensor_data: The state observed.
+
+        Returns
+        ----------
+            NumPy array of float32: Normalized Sensor Data
+        """
+
+        sensor_data = np.array(sensor_data) / MAX_SENSOR_VALUE
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py]: It's a copy of the argument, I recognized my mistake and I fixed it

File: webots/controllers/RL_Supervisor/agent.py [hunk 19]
```diff
@@ -0,0 +1,544 @@
...
+            self.action = action_mean.numpy()[0]
+
+        return self.action
+
+    def send_motor_speeds(self, state):
+        """Sends the motor speeds to the robot."""
+
+        pre_action = self.predict_action(state)
+
+        # Get motor speed difference
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py:237]: I have already added comment to clarify what pre_action contains

File: webots/controllers/RL_Supervisor/agent.py [hunk 20]
```diff
@@ -0,0 +1,507 @@
...
+        # determines if the robot turns left or right.
+        left_motor_speed = int(self.__top_speed - speed_difference)
+        right_motor_speed = int(self.__top_speed + speed_difference)
+
+        control_data = struct.pack("2H", left_motor_speed, right_motor_speed)
+        self.data_sent = self.__serialmux.send_data("SPEED_SET", control_data)
+
+        # Failed to send data. Appends the data to unsent_data List.
+        if self.data_sent is False:
+            self.unsent_data.append(("SPEED_SET", control_data))
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py]: In every timestep, I check in the supervisor whether any data has not been sent. If that's the case, I use the retry_unsent_data function in the supervisor at line 174 to resend the data

File: webots/controllers/RL_Supervisor/plotting.py [hunk 21]
```diff
@@ -0,0 +1,47 @@
+""" Plotting script  with Matplotlib """
+
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/plotting.py:2]: I have added a description below the line to clarify its purpose and functionality.

File: webots/controllers/RL_Supervisor/agent.py [hunk 22]
```diff
@@ -0,0 +1,507 @@
...
+#
+
+
+################################################################################
+# Imports
+################################################################################
+import csv
+import os
+import struct
+import numpy as np  # pylint: disable=import-error
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/agent.py]: I tried without disable and pylint give me an import error in Python version 3.9 and 3.10 when I pushed my work.

File: webots/controllers/RL_Supervisor/networks.py [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+
+################################################################################
+# Imports
+################################################################################
+
+from tensorflow import keras  # pylint: disable=import-error
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/networks.py]: I tried without disable and pylint give me an import error in Python version 3.9 and 3.10 when I pushed my work.

File: webots/controllers/RL_Supervisor/trajectory_buffer.py (hunk 11 above)
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/trajectory_buffer.py]: I tried without disable and pylint give me an import error in Python version 3.9 and 3.10 when I pushed my work.

File: lib/APPReinforcementLearning/src/App.cpp [hunk 24]
```diff
@@ -149,16 +151,30 @@ void App::loop()
...
             m_modeSelectionSent = true;
         }
     }
 
-    if (false == isDataSent)
+    if (false == isSensorDataSent)
     {
-        /* Failed to send data to the supervisor. Go to error state. */
-        ErrorState::getInstance().setErrorMsg("DSF");
+        /* Failed to send Senssor data to the supervisor. Go to error state. */
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:162]: This works. But could be done better by using if-else blocks

File: webots/controllers/RL_Supervisor/trajectory_buffer.py (hunk 13 above)
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/trajectory_buffer.py:43]: resolved with issue #156

File: webots/controllers/RL_Supervisor/networks.py (hunk 12 above)
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/networks.py]: resolved with issue #156

File: lib/APPReinforcementLearning/src/App.cpp [hunk 25]
```diff
@@ -123,35 +125,56 @@ void App::loop()
             payload = {SMPChannelPayload::Status::DONE};
         }
 
-        /* Ignoring return value, as error handling is not available. */
-        (void)m_smpServer.sendData(m_serialMuxProtChannelIdStatus, &payload, sizeof(payload));
+        isDataSent = m_smpServer.sendData(m_serialMuxProtChannelIdStatus, &payload, sizeof(payload));
```
Comment (gabryelreyes) [inline lib/APPReinforcementLearning/src/App.cpp:128]: This is too dificult to read. Why not try something like this?
```suggestion
        if (false == m_smpServer.sendData(m_serialMuxProtChannelIdStatus, &payload, sizeof(payload)))
{
    /* Failed to send Status data to the Supervisor. Go to error state. */
    ErrorState::getInstance().setErrorMsg("SD_SF");
    m_systemStateMachine.setState(&ErrorState::getInstance());
}
else
{
    m_statusTimer.restart();
}
```

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: webots/controllers/RL_Supervisor/requirements.txt [hunk 26]
```diff
@@ -0,0 +1,5 @@
+tensorflow
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/requirements.txt:1]: Not all dependencies are listed

File: webots/controllers/RL_Supervisor/requirements.txt [hunk 27]
```diff
@@ -0,0 +1,5 @@
...
+tensorflow-probability
+tf-keras
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/requirements.txt:3]: Add specific versions to each dependency

File: webots/controllers/RL_Supervisor/requirements.txt [hunk 28]
```diff
@@ -0,0 +1,5 @@
...
+
```
Comment (gabryelreyes) [inline webots/controllers/RL_Supervisor/requirements.txt:4]: remove empty line

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

**Review by gabryelreyes — CHANGES_REQUESTED** [review]

File: webots/controllers/RL_Supervisor/requirements.txt [hunk 29]
```diff
@@ -0,0 +1,4 @@
+numpy==1.26.4
+tensorflow==2.10.0
+tensorflow-probability==0.15.0
+git+https://github.com/gabryelreyes/SerialMuxProt.git#egg=SerialMuxProt&subdirectory=python/SerialMuxProt
```
[the remark this answers is not in the document]
Comment (akrambzeo) [reply webots/controllers/RL_Supervisor/requirements.txt:4]: Keras is integrated into TensorFlow

**Review by gabryelreyes — APPROVED** [review]
